### ISSUE TITLE
[Linux] Segmentation Fault when trying to start the game

### ISSUE THREAD

**Issue by Enygma2002** [ValveSoftware/Dota-2#1250]
Since very recently (~12th of June), as far as I can tell, after one of the recent updates, Dota2 fails to launch on my Fedora 25 machine. Was working fine before.

Trying to launch the game, I see the loading screen (black screen with Dota logo in the center), it takes about 5-10 seconds and the game crashes with a `Segmentation Fault`.

**Console log:**
```
GameAction [AppID 570, ActionID 1] : RunGame waiting for user response to LaunchOption ""
GameAction[AppID 570, ActionID 1] : RunGame continues with user response "0"
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to Starting with ""
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to UpdatingAppInfo with ""
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to SynchronizingCloud with ""
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to CreatingProcess with ""
GameAction [AppID 570, ActionID 2] : LaunchApp waiting for user response to CreatingProcess ""
GameAction[AppID 570, ActionID 2] : LaunchApp continues with user response "CreatingProcess"
Game update: AppID 570 "Dota 2", ProcID 3182, IP 0.0.0.0:0
>>> Adding process 3182 for game ID 570
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to WaitingGameWindow with ""
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to Completed with ""
Generating new string page texture 625: 128x256, total string texture memory is 1.88 MB
>>> Adding process 3183 for game ID 570
pid 3191 != 3183, skipping destruction (fork without exec?)
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
>>> Adding process 3192 for game ID 570
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
Using breakpad crash handler
Game update: AppID 570 "Dota 2", ProcID 3193, IP 0.0.0.0:0
Setting breakpad minidump AppID = 570
Forcing breakpad minidump interfaces to load
Looking up breakpad interfaces from steamclient
Calling BreakpadMiniDumpSystemInit
Looking up breakpad interfaces from steamclient
Calling BreakpadMiniDumpSystemInit
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561198001206110 [API loaded yes]
Steam_SetMinidumpSteamID:  Setting Steam ID:  76561198001206110
Setting breakpad minidump AppID = 373300
>>> Adding process 3193 for game ID 570
ATTENTION: default value of option vblank_mode overridden by environment.
ATTENTION: default value of option vblank_mode overridden by environment.
Attempting to create OpenGL context: v3.3 (Core profile)...success.
Attempting to create OpenGL context: v3.3 (Core profile)...success.
ATTENTION: default value of option vblank_mode overridden by environment.
ATTENTION: default value of option vblank_mode overridden by environment.
Installing breakpad exception handler for appid(gameoverlayui)/version(20170608043310)
Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)
Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)
DumpContextInfo: OpenGL vendor NVIDIA Corporation
DumpContextInfo: OpenGL renderer GeForce GTX 960M/PCIe/SSE2
DumpContextInfo: Using OpenGL context version 3.3
DumpContextInfo: Context supports GLSL version 3.30 NVIDIA via Cg compiler
Fontconfig warning: line 38: unknown element "fontpattern"
Fontconfig warning: line 39: unknown element "fontpattern"
Fontconfig warning: line 40: unknown element "fontpattern"
Fontconfig warning: line 41: unknown element "fontpattern"
Fontconfig warning: line 42: unknown element "fontpattern"
Fontconfig warning: line 43: unknown element "fontpattern"
Fontconfig warning: line 44: unknown element "fontpattern"
Fontconfig warning: line 45: unknown element "fontpattern"
Fontconfig warning: line 46: unknown element "fontpattern"
Fontconfig warning: line 47: unknown element "fontpattern"
Fontconfig warning: line 48: unknown element "fontpattern"
Fontconfig warning: line 49: unknown element "fontpattern"
Fontconfig warning: line 50: unknown element "fontpattern"
Fontconfig warning: line 51: unknown element "fontpattern"
Fontconfig warning: line 52: unknown element "fontpattern"
Fontconfig warning: line 53: unknown element "fontpattern"
Fontconfig warning: line 54: unknown element "fontpattern"
Fontconfig warning: line 55: unknown element "fontpattern"
Fontconfig warning: line 56: unknown element "fontpattern"
Fontconfig warning: line 57: unknown element "fontpattern"
Fontconfig warning: line 58: unknown element "fontpattern"
Fontconfig warning: line 59: unknown element "fontpattern"
Fontconfig warning: line 60: unknown element "fontpattern"
Fontconfig warning: line 61: unknown element "fontpattern"
Fontconfig warning: line 62: unknown element "fontpattern"
Fontconfig warning: line 63: unknown element "fontpattern"
Fontconfig warning: line 64: unknown element "fontpattern"
Fontconfig warning: line 65: unknown element "fontpattern"
Installing breakpad exception handler for appid(gameoverlayui)/version(1.0)
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
>>> Adding process 3223 for game ID 570
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_64/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
crash_20170614010441_1.dmp[3229]: Uploading dump (out-of-process)
/tmp/dumps/crash_20170614010441_1.dmp
>>> Adding process 3227 for game ID 570
crash_20170614010441_1.dmp[3229]: Finished uploading minidump (out-of-process): success = yes
crash_20170614010441_1.dmp[3229]: response: Discarded=1
crash_20170614010441_1.dmp[3229]: file ''/tmp/dumps/crash_20170614010441_1.dmp'', upload yes: ''Discarded=1''
pid 3229 != 3227, skipping destruction (fork without exec?)
/home/Enygma/.local/share/Steam/steamapps/common/dota 2 beta/game/dota.sh: line 104:  3193 Segmentation fault      (core dumped) ${STEAM_RUNTIME_PREFIX} ${GAME_DEBUGGER} "${GAMEROOT}"/${GAMEEXE} "$@"
Game removed: AppID 570 "Dota 2", ProcID 3193
No cached sticky mapping in ActivateActionSet.[0614/010926.276563:ERROR:web_plugin_impl.cc(38)] Widevine registration is not supported after context initialization
```

**System info:**
```
Computer Information:
    Manufacturer:  Unknown
    Model:  Unknown
    Form Factor: Laptop
    No Touch Input Detected
    
Processor Information:
    CPU Vendor:  GenuineIntel
    CPU Brand:  Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz
    CPU Family:  0x6
    CPU Model:  0x3c
    CPU Stepping:  0x3
    CPU Type:  0x0
    Speed:  3600 Mhz
    8 logical processors
    4 physical processors
    HyperThreading:  Supported
    FCMOV:  Supported
    SSE2:  Supported
    SSE3:  Supported
    SSSE3:  Supported
    SSE4a:  Unsupported
    SSE41:  Supported
    SSE42:  Supported
    AES:  Supported
    AVX:  Supported
    CMPXCHG16B:  Supported
    LAHF/SAHF:  Supported
    PrefetchW:  Unsupported
    
Network Information:
    Network Speed:  
    
Operating System Version:
    "Fedora release 25 (Twenty Five)" (64 bit)
    Kernel Name:  Linux
    Kernel Version:  4.10.17-200.fc25.x86_64
    X Server Vendor:  Fedora Project
    X Server Release:  11903000
    X Window Manager:  KWin
    Steam Runtime Version:  steam-runtime-beta-release_2016-12-05
    
Video Card:
    Driver:  Intel Open Source Technology Center Mesa DRI Intel(R) Haswell Mobile 

    Driver Version:  3.0 Mesa 17.0.5
    OpenGL Version: 3.0
    Desktop Color Depth: 24 bits per pixel
    Monitor Refresh Rate: 60 Hz
    VendorID:  0x8086
    DeviceID:  0x416
    Revision Not Detected
    Number of Monitors:  1
    Number of Logical Video Cards:  1
    Primary Display Resolution:  1920 x 1080
    Desktop Resolution: 1920 x 1080
    Primary Display Size: 13.78" x 7.48"  (15.67" diag)
                                            35.0cm x 19.0cm  (39.8cm diag)
    Primary VRAM Not Detected
    
Sound card:
    Audio device: Intel Haswell HDMI
    
Memory:
    RAM:  15958 Mb
    
Miscellaneous:
    UI Language:  English
    LANG:  en_US.UTF-8
    Microphone:  Not set
    Steam Controller Cable and Base:  Not set
    Total Hard Disk Space Available:  185121 Mb
    Largest Free Hard Disk Block:  27863 Mb
    VR Headset: None detected
    
Recent Failure Reports:
    Tue Jun 13 22:03:48 2017 GMT: file ''/tmp/dumps/crash_20170614010338_1.dmp'', upload yes: ''Discarded=1''
    Tue Jun 13 22:04:51 2017 GMT: file ''/tmp/dumps/crash_20170614010441_1.dmp'', upload yes: ''Discarded=1''
```

**Comment by gdrewb-valve** [comment]
Have you validated your Dota install?

**Comment by Enygma2002** [comment]
@gdrewb-valve Thanks for the reply.

I've just tried a validation and no luck.

Some extra information that may be useful:
While showing the loading screen, about 1/4 of the way, the mouse pointer turns from the system one to the Dota 2 specific one (bigger/chunkier one). Maybe that helps you better identify the point of failure.

Here's the crash dump file, if it helps debugging:
http://s000.tinyupload.com/index.php?file_id=05648783704226240929 (~454KB in size)

**Comment by kisak-valve** [comment]
For reference, the attached minidump is a SIGSEGV in libengine2.so (part of Dota 2).

**Comment by ghost** [comment]
Get ubuntu.

**Comment by gdrewb-valve** [comment]
The larger team is looking at crash issues that affect all platforms and there won't be a lot of immediate status here.

**Comment by Enygma2002** [comment]
Hoping the priority is high, since right now I can't start the game at all, on Fedora. Not sure if it's only about my install or if it reproduces on any recent Fedora machine.

Thanks for the feedback so far.

**Comment by gdrewb-valve** [comment]
Unfortunately it's impossible to say when this will get addressed.  The larger team is going through crashes by the frequency they occur overall so hopefully yours matches up with something.

**Comment by gdrewb-valve** [comment]
You might also want to check on forums for your distro to see if there are people with similar problems.  As you're running without the Steam runtime there may be some system conflict that's popping up.

**Comment by Enygma2002** [comment]
@gdrewb-valve:
> As you're running without the Steam runtime

Could you please clarify what you mean by that, how did you determine that and how can I try to run "with" the Steam runtime?

My Steam install is standard (reinstalled about 1 month ago). The game was running fine until the recent updates.

**Side note**:
This is the second bug I`m reporting and every time I mention that I'm running it on Fedora, I am referred to look for solutions or other people with the same problem on Fedora-related forums, because all the Dota devs seem to be running Ubuntu. It's not such a nice feeling to be shunned out that way, specially since:
1) Fedora is one of the major Linux distributions and Dota 2 support says **SteamOS + Linux**, not "SteamOS + Ubuntu**. Feels like everything other than Ubuntu is a second class citizen.
2) A Dota 2 event is currently running, for which I have purchased a **Battle Pass** / Compendium which I can't currently utilize and which will eventually expire. Random crashes are one thing, but consistent crashes that make opening the game impossible deserve to be handled with priority, don't you agree? I wouldn't want to end up missing out on the event due to a recent update (i.e. not my fault).

Thanks for the good job you guys are doing and I'm hoping someone can be pinged to at least understand the nature of the problem. I'm available to provide any extra information needed to diagnose the problem.

**Comment by Enygma2002** [comment]
Did not really find much related to Fedora.

The only recent relevant thing I found (june 2017) was on the opensuse forum: https://forums.opensuse.org/showthread.php/525175-After-Steam-update-the-games-is-not-launching?p=2825138#post2825138

However, the proposed workaround `STEAM_RUNTIME_PREFER_HOST_LIBRARIES=0 steam` did not work for me:
```
Game update: AppID 570 "Dota 2", ProcID 10959, IP 0.0.0.0:0
>>> Adding process 10959 for game ID 570
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to WaitingGameWindow with ""
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/Enygma/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
/bin/bash: relocation error: /usr/lib64/primus/libGL.so.1: symbol _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEED1Ev, version GLIBCXX_3.4.21 not defined in file libstdc++.so.6 with link time reference
GameAction [AppID 570, ActionID 2] : LaunchApp changed task to Completed with ""
>>> Adding process 10960 for game ID 570

(steam:10837): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.

... repeated numerous times ...

(steam:10837): LIBDBUSMENU-GLIB-WARNING **: Trying to remove a child that doesn't believe we're it's parent.
Game removed: AppID 570 "Dota 2", ProcID 10959 
```

**Comment by kisak-valve** [comment]
Your log snippet indicates that you are using an optimus setup, which means you may be encountering something similar to https://github.com/ValveSoftware/steam-for-linux/issues/4822.

What launch options are you using with Dota 2? Does running the game on the Intel gpu or adding `LD_PRELOAD="libpthread.so.0 libGL.so.1" primusrun %command%` to the launch options improve the situation?

**Comment by Enygma2002** [comment]
Found the culprit!

My launch options were `vblank_mode=0 primusrun %command% -consolelog -conclearlog +sys_info`, most likely from previous attempts to debug a previous problem (https://github.com/ValveSoftware/Dota-2/issues/1195). For some reason, those launch options seemed to work fine up until the recent updates.

The solution seems to be to simply revert to my standard launch options: `vblank_mode=0 primusrun %command%` which successfully launches the game.

Thanks, @kisak-valve, for the inspiration. I did not think to check the launch options.

I`m now able to enjoy Dota again :)
